Bgp control plane:  add route aggregation feature

romanspb80 opened the pull request:
Add route aggregation feature in BGP Control Plane (RFC 4632)
New route aggregation options under the CiliumBGPAdvertisement.
```
apiVersion: cilium.io/v2alpha1
kind: CiliumBGPAdvertisement
metadata:
  name: bgp-advertisements
  labels:
    advertise: bgp
spec:
  advertisements:
    - advertisementType: "Service"
      service:
        aggregationLengthIPv4: 24                  <<<<<<<<<<<<<<   New Field
        aggregationLengthIPv6: 120                 <<<<<<<<<<<<<<   New Field
```


Fixes: #36777

Signed-off-by: Roman Ptitcyn <romanspb@yahoo.com>

harsimran-pabla reviewed (changes requested):
Hi @romanspb80 Before jumping into the code review of the implementation, I think we need to agree on the API change. I still think user should specify aggregation length instead of dynamically calculating it based on the boolean flag.

In vast majority of cases there is going to be single IP returned from [getServicePrefix](https://github.com/cilium/cilium/pull/37275/files#diff-96a914855fa998ca55f8d09877045422300b7abf80ec4d39a4340b0840b935e0R534)
, so with single /32 or /128 prefix there will be nothing to aggregate. 

Have you tested this change on kubernetes cluster with sample loadbalancer service ? can you provide some example outputs how aggregation will look like?

harsimran-pabla commented on pkg/k8s/apis/cilium.io/v2alpha1/bgp_advert_types.go:
```diff
@@ -122,6 +122,16 @@ type BGPAdvertisement struct {
 // BGPServiceOptions defines the configuration for Service advertisement type.
 type BGPServiceOptions struct {
+	// IPv4 mask to aggregate BGP route advertisements of service
+	// +kubebuilder:validation:Minimum=24
```
I don't think we need to assume min value to be 24 or 120 in case of v6. We can skip the min in both v4 and v6 cases. 

Max values makes sense to me.

harsimran-pabla commented on pkg/bgpv1/manager/reconcilerv2/service.go:
```diff
@@ -535,14 +535,21 @@ func (r *ServiceReconciler) getServiceAFPaths(p ReconcileParams, desiredPeerAdve
-
 					// we only add path corresponding to the family of the prefix.
 					if agentFamily.Afi == types.AfiIPv4 && prefix.Addr().Is4() {
+						if advert.Service.AggregationLengthIPv4 != nil {
```
We should check for eTP=Local here and omit prefix aggregation for such services.

harsimran-pabla commented on pkg/bgpv1/manager/reconcilerv2/service.go:
```diff
@@ -706,12 +713,19 @@ func (r *ServiceReconciler) getLoadBalancerIPRoutePolicy(p ReconcileParams, peer
 		}
 
+		var mask int
+		if advert.Service.AggregationLengthIPv4 != nil && svc.Spec.ExternalTrafficPolicy != slim_corev1.ServiceExternalTrafficPolicyLocal {
```
Logic seems to be checking only for IPv4. Don't we need similar checks for IPv6?

harsimran-pabla reviewed (changes requested):
Thanks @romanspb80 for the changes. API looks good to me, there are few other comments regarding the implementation.

- Route policies should be done for all service types ( external IP, cluster IP ), currently it implemented for LB Ingress IP only.
- Unit tests are missing.
- Route policies which are merged [here](https://github.com/cilium/cilium/pull/37275/files#diff-96a914855fa998ca55f8d09877045422300b7abf80ec4d39a4340b0840b935e0R283) should be sorted based on prefix length. For eg, if multiple advertisements match same service, longer prefix should be first.

tommyp1ckles reviewed (approved):
@cilium/sig-k8s CRD changes look good to me. I do see that @harsimran-pabla had a comment regarding the need for the min values.

harsimran-pabla commented on pkg/bgpv1/manager/reconcilerv2/service.go:
```diff
@@ -702,19 +710,35 @@ func (r *ServiceReconciler) getLoadBalancerIPRoutePolicy(p ReconcileParams, peer
+
 		if family.Afi == types.AfiIPv4 && addr.Is4() {
-			v4Prefixes = append(v4Prefixes, &types.RoutePolicyPrefixMatch{CIDR: netip.PrefixFrom(addr, addr.BitLen()), PrefixLenMin: addr.BitLen(), PrefixLenMax: addr.BitLen()})
+			if advert.Service.AggregationLengthIPv4 != nil && svc.Spec.ExternalTrafficPolicy != slim_corev1.ServiceExternalTrafficPolicyLocal {
```
Consider moving this out into a helper method. Same code is repeated 3 times.

romanspb80 replied:
done

harsimran-pabla commented on pkg/bgpv1/manager/reconcilerv2/service.go:
```diff
@@ -702,19 +710,35 @@ func (r *ServiceReconciler) getLoadBalancerIPRoutePolicy(p ReconcileParams, peer
 		return nil, nil
 	}
 
+	sort.SliceStable(v4Prefixes, func(i, j int) bool {
```
This sorting here is not what I had in mind. Case is something like this 
- Advert A picks Srv X and configures aggregate prefix = 192.168.100.0/24 with community attr 65000:100
- Advert B picks Srv X as well, and configures exact prefix = 192.168.100.1/32 with community attr 65000:200

Route policy created by these two advertisements are going to have same [name](https://github.com/cilium/cilium/blob/main/pkg/bgpv1/manager/reconcilerv2/service.go#L718). 

So, when route policies are getting merged [here](https://github.com/cilium/cilium/blob/main/pkg/bgpv1/manager/reconcilerv2/service.go#L262-L315), we will just add statements in random order. Consequence of this is that if statement with `192.168.100.0/24` prefix is first then BGP announcement for `192.168.100.1` IP will go out with community 65000:100, instead of 65000:200.

I think we can sort the statements in route-policy after merging two policies [here](https://github.com/cilium/cilium/blob/main/pkg/bgpv1/manager/reconcilerv2/policies.go#L252).

romanspb80 replied:
thx. got it. will do

romanspb80 replied:
done.
But I have doubts about the necessity of this sorting. Because the target (peering) BGP-router will make route selection based on prefix length and BGP attributes priority,  not on the order of the received advertisements.

harsimran-pabla replied:
That's right - target router is going to make routing based on best path selection. It is mostly necessary for traffic engineering in core network based on BGP attributes. For eg, /32 routes might not be allowed to leave a region vs /24 summary route is allowed.

harsimran-pabla commented on pkg/bgpv1/manager/reconcilerv2/policies.go:
```diff
@@ -249,6 +249,16 @@ func MergeRoutePolicies(policyA *types.RoutePolicy, policyB *types.RoutePolicy)
 		}
 	}
 
+	for n := range mergedPolicy.Statements {
```
nit: can you please add a comment describing why sorting is done for future reference.

romanspb80 replied:
done

harsimran-pabla reviewed (commented):
Thanks for the latest changes, it is looking good to me - couple of more things
- please update the PR description with latest way of configuring summarization.
- Bump the CRD version [here](https://github.com/cilium/cilium/blob/main/pkg/k8s/apis/cilium.io/register.go#L18), since  you are changing the CRD.
- Update the documentation with example on how to use this feature. Maybe add a section under [service advertisements](https://github.com/cilium/cilium/blob/main/Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst#service-virtual-ips)

harsimran-pabla reviewed (approved):
Thanks @romanspb80 for your contribution and patiently working with me. This is very useful feature. 🚀

romanspb80 committed:
Add route aggregation feature in BGP Control Plane (RFC 4632)

New route aggregation option under the CiliumBGPAdvertisement.
apiVersion: cilium.io/v2alpha1
kind: CiliumBGPAdvertisement
metadata:
  name: bgp-advertisements
  labels:
    advertise: bgp
spec:
  advertisements:
    - advertisementType: "Service"
      service:
        aggregationLengthIPv4: 24                   <<<<<<<<<<<<<<   New Field
        aggregationLengthIPv6: 120                  <<<<<<<<<<<<<<   New Field

Fixes: #36777

Signed-off-by: Roman Ptitcyn <romanspb@yahoo.com>
